src/OTP.hs

```diff
@@ -0,0 +1,365 @@
+{-# LANGUAGE DerivingStrategies #-}
+
+-- | Implements HMAC-Based One-Time Password Algorithm as defined in RFC 4226 and
+--  Time-Based One-Time Password Algorithm as defined in RFC 6238.
+--
+-- Only HMAC-SHA-256 and HMAC-SHA-512 are provided.
+module OTP
+  ( -- * HOTP
+    hotp256
+  , hotp256Check
+  , hotp512
+  , hotp512Check
+
+    -- * TOTP
+  , totp256
+  , totp256Check
+  , totp512
+  , totp512Check
+  --
+  --   -- * Auxiliary
+  , totpCounter
+  , counterRange
+  , totpCounterRange
+  ) where
+
+import Data.Bits
+import Data.ByteString qualified as BS
+import Data.Serialize.Get
+import Data.Serialize.Put
+import Data.Time
+import Data.Time.Clock.POSIX
+import Data.Word
+import Sel.HMAC.SHA256 qualified as SHA256
+import Sel.HMAC.SHA512 qualified as SHA512
+
+-- | Compute HMAC-Based One-Time Password using secret key and counter value.
+--
+-- >>> hotp "1234" 100 6
+-- 317569
+--
+-- >>> hotp SHA512 "1234" 100 6
+-- 134131
+--
+-- >>> hotp SHA512 "1234" 100 8
+-- 55134131
+hotp256
+  :: SHA256.AuthenticationKey
+  -- ^ Shared secret
+  -> Word64
+  -- ^ Counter value
+  -> Word
+  -- ^ Number of digits in a password. Should be between 6 and 8.
+  -> IO Word32
+  -- ^ HOTP
+hotp256 key counter digits = do
+  let msg = runPut $ putWord64be counter
+  hash <- SHA256.authenticationTagToBinary <$> SHA256.authenticate msg key
+  let w = truncateHash $ BS.unpack hash
+  pure $ w `mod` (10 ^ digits)
+
+hotp512
+  :: SHA512.AuthenticationKey
+  -- ^ Shared secret
+  -> Word64
+  -- ^ Counter value
+  -> Word
+  -- ^ Number of digits in a password
+  -> IO Word32
+  -- ^ HOTP
+hotp512 key counter digits = do
+  let msg = runPut $ putWord64be counter
+  hash <- SHA512.authenticationTagToBinary <$> SHA512.authenticate msg key
+  let w = truncateHash $ BS.unpack hash
+  pure $ w `mod` (10 ^ digits)
+
+truncateHash :: [Word8] -> Word32
+truncateHash b =
+  let offset = last b .&. 15 -- take low 4 bits of last byte
+      rb = BS.pack $ take 4 $ drop (fromIntegral offset) b -- resulting 4 byte value
+   in case runGet getWord32be rb of
+        Left e -> error e
+        Right res -> res .&. (0x80000000 - 1) -- reset highest bit
```

Comment:
Why not just hardcode the mask instead of doing arithmetic on hex? Also, this forces a clear on the highest bit, which isn't really 'resetting' anything.

Reply:
good question you'll have to ask hvr :P

Reply:
modified the comment